include/openmc/shared_array.h

```diff
@@ -0,0 +1,119 @@
+#ifndef OPENMC_SHARED_ARRAY_H
+#define OPENMC_SHARED_ARRAY_H
+
+//! \file shared_array.h
+//! \brief Shared array data structure
+
+#include<memory>
+
+
+namespace openmc {
+
+//==============================================================================
+// Class declarations
+//==============================================================================
+
+// The SharedArray is an array that is capable of being appended to in an
+// thread safe manner by use of atomics. It only provides protection for the
+// use cases currently present in OpenMC. Namely, it covers the scenario where
+// multiple threads are appending to an array, but no threads are reading from
+// or operating on it in any other way at the same time. Multiple threads can
+// call the thread_safe_append() function concurrently and store data to the
+// object at the index returned from thread_safe_append() safely, but no other
+// operations are protected.
+template <typename T> 
+class SharedArray { 
```

Comment:
I would prefer we have SharedArray stick as close as possible to the semantics of an STL sequence container. Instead of
```C++
   int64_t thread_safe_append()
```
it could be
```C++
  int64_t push_back(const T&)
```

Then we can add a `QueueItem(const Particle&, int64_t index)` constructor and append to the queues with `queue.push_back({p, buffer_indx})`.

Reply:
@jtramm and I did discuss this a bit. I was also of the opinion that we should name it `push_back`, but it does give the false impression that you can just use this class like a vector (when in fact there's a hard-limit on the size; it's not dynamic). I'm fine with the name either way. Having the method take a `T` argument does make sense though.

Reply:
This idea makes sense if we were building a more generic SharedArray class, and definitely would make it much more similar to the STL. However, there are some downsides to that:

1 - As @paulromano says, providing a push_back() function with the same arguments as an STL vector could be misleading. Vectors are commonly used by initializing them as being empty, then calling push_back() for all elements that need to be added, which is not what this function is providing since the SharedArray must be allocated to a fixed size. So, I think that calling it `thread_safe_append()` has two nice benefits. The first is that it means people will not confuse its functionality with that of the C++ STL vector's `push_back()`. The second is that it specifically calls attention to the fact that this is the only method in the class that is threadsafe -- all other member functions are not protected.

2 -  As far as the switch to an interface where the caller passes in a reference to the actual object to be appended (e.g., `thread_safe_append(const T&)`), there are some complications to that. Currently, the SharedArray is used in two places. The first is as a queue for the event-based mode (where this would work just fine), and the second is as the shared fission bank [here](https://github.com/openmc-dev/openmc/blob/d24667f348a8b728196c95c50dd438639a28fcda/src/physics.cpp#L181-L223). When fission sites are created, it is helpful for the caller to know if the fission bank is full or not, as if it is full a lot of extra work of actually sampling the fission site can be skipped. The current `thread_safe_append()` function allows for this functionality, as it will return -1 if it is full without requiring that the user have the final element that they want to add before checking. When creating fission sites, we could alternatively add a different method that atomically checks the size of the fission bank first, then calls a different `thread_safe_append(const T&)` function later on, but for the average use case (when the fission bank is not full) this would be doubling the number of atomic operations that we are performing here. We would also have to add in some special logic to handle the cases where the fission bank runs out of space between the time we query its size, and when we call the `thread_safe_append(const T&)` function, which in my opinion would make things messier. One way around this issue would to provide two different methods for appending. In the first, we keep the current `thread_safe_append()` method for use when creating fission sites, and a second new method `thread_safe_append(const T&)` for use in the event-based queuing logic where we don't care about checking if there is space or not. My thought is that this is sort complicating things though, as we are providing two different methods to accomplish the same task.

Overall, I think that the SharedArray append interface as it is now is tailored nicely to the current use cases in OpenMC. Changing the interface and/or naming would introduce some additional complexities that I don't think would improve the code. I think this is unfortunately part of the territory of writing a thread safe class -- namely, achieving optimal performance and providing an elegant interface can only be accomplished for a specific use case, rather than a thread safe catch all. If we wanted all of our functions to be thread safe useable for all use cases, but still offer an STL-like interface, we would have to add in a lot more atomic operations and require a lot of extra logic on the calling side, which I don't think makes sense. In our case, I would rather keep the class as simple as possible and tailored to our specific use cases rather than trying to optimally mimic the STL interfaces.

Reply:
I don't mind having `create_fission_sites` perform unused work when the bank is full so that we can have this function accept a `T` argument. A full fission bank is a rare case, and it implies a host of important issues so a small performance drop there is nbd. Does a `thread_safe_append(const T&)` make sense for `create_fission_sites` if we aren't worried about that small cost? Fair points on the function name.

Reply:
I've updated it a bit to hopefully get a best of both worlds option. The function now has the prototype:

```C++
int64_t thread_safe_append(const T& value = {})
```

which allows for more intuitive appending by directly passing an element, while still also allowing us to pass nothing and skip logic in the `create_fission_sites` functions by functioning the same way as the original function if no argument is given by the caller.

I also added in the constructors to `EventQueueItem` to make it work so that in our event-based queuing routines we can now append e.g.:

```C++
queue.thread_safe_append({p, buffer_idx});
```

Let me know if you like this new version, or if you would still prefer to go with something like:

```C++
int64_t thread_safe_append(const T& value)
```

which will involve some reworking of logic in the physics.cpp and physics_mg.cpp `create_fission_sites` to make work.

Reply:
Oops, I approved the changes, but I just realized that a `value = {}` approach will probably result in two writes to the fission bank every time we bank a particle which might have a performance impact. It's not worth spending much time on so I'll leave it up to you if you want to just revert back to 958a77f or go forward with adjusting the `create_fission_sites` logic. Apologies for the back and forth.

Reply:
True! I will go the ```int64_t thread_safe_append(const T& value)``` route.